R/cube-result.R

```diff
@@ -434,3 +402,147 @@ setMethod("bases", "CrunchCube", function (x, margin=NULL) {
+#' correctly compute margins and percentages off of them.
+#'
+#' These functions also generalize to MultitableResults and TabBookResults,
+#' which are returned from a [tabBook()] request. When called on one of those
+#' objects, they effectively apply over each CrunchCube contained in them.
+#'
+#' `bases` is an additional method for CrunchCubes. When making weighted
+#' requests, `bases` allows you to access the unweighted counts for every cell
+#' in the resulting table (array). The `bases` function takes a "margin"
+#' argument to work like `margin.table`, or with `margin=0` gives all cell
+#' counts.
+#'
+#' `collapse.dimensions` returns a cube that collapses the dimensions given in
+#' `margin`. This is useful if you want to get counts that are equivalent to a
+#' univariate cube from a multivariate cube. For example
+#' `collapse.dimensions(crtabs(~ fruit + pets, ds), 2)` will be equal to
+#' `crtabs(~ fruit, ds)` and `collapse.dimensions(crtabs(~ fruit + pets, ds),
+#' 1)` will be equal to `crtabs(~ pets, ds)`.
+#'
+#' @param x a CrunchCube
+#' @param margin index, or vector of indices to generate margin for. See
+#'   [base::prop.table()]. `bases()` accepts `0` as an additional valid value
+#'   for `margin`, which yields the unweighted counts for the query.
+#' @param digits For `round`, the number of decimal places to round to. See
+#'   [base::round()]
+#'
+#' @return When called on CrunchCubes, these functions return an `array`.
+#'   Calling prop.table on a MultitableResult returns a list of prop.tables of
+#'   the CrunchCubes it contains. Likewise, prop.table on a TabBookResult
+#'   returns a list of lists of prop.tables.
+#' @name cube-computing
+#' @aliases cube-computing margin.table prop.table bases round
+#'   collapse.dimensions
+#' @seealso [base::margin.table()] [base::prop.table()]
+NULL
+
+#' @rdname cube-computing
+#' @export
+setMethod("margin.table", "CrunchCube", function (x, margin=NULL) {
+    cubeMarginTable(x, margin)
+})
+
+#' Convert from user margins to real cube margins or vice versa
+#'
+#' It is helpful to programmatically move from user-specified margins to real
+#' cube margins that apply to the higher-dimensional real cube (with
+#' `user2real`). Or to move from the higher-dimensional real cube to user cube
+#' (with `real2user`).
+#'
+#' @param margin the margin or margins for the user cube to be translated
+#' @param dimTypes dimension types from `getDimTypes()` (by default:
+#'   `getDimTypes(cube)`)
+#' @param cube the cube to translate the margin for (optional if `dimTypes` is
+#'   explicitly supplied)
+#' @param dedupe logical, should the user dimensions in the result be
+#'   deduplicated (for real2user only)
+#'
+#' @return margin or margins in the higher-dimension real cube
+#'
+#' @keywords internal
+#'
+#' @name margin-translation
+NULL
+
+#' @rdname margin-translation
+user2real <- function(margin, dimTypes = getDimTypes(cube), cube) {
```

Comment:
These function names are too ambiguous. They should say "margin" in them, I think.